ENH: Add qv photochemistry diagnostics

DWesl opened the pull request:
Closes #255 

Adds variables to accumulate water vapor photochemical tendencies from photochemistry, document them in the metadata files (for code generation?), and add code to the photochemistry module to calculate the tendencies due to each term of the parametrization, based on #196.

The resulting diagnostics are `dtend_qv_photochem`, `dtend_qv_prodloss`, and `dtend_qv_o3mix`, since `dtend_o3_o3mix` is the analogous term in the ozone photochemistry equation and I don't know how to change that name.  This represents the $\left.\frac{d(P-L)}{dr}\right|_0 (r - r_0) = -\frac{1}{\tau} (r - r_0)$ term, implementing the dependence on mixing ratio/mass fraction or relaxation to climatology, depending on interpretation.

DWesl committed:
ENH: Add water vapor photochemistry diagnostics.

DWesl committed:
Merge branch 'add-qv-photochemistry-diagnostics' of global-workflow-20250128/sorc/ufs_model.fd/FV3/ccpp/physics into HEAD

DWesl committed:
Merge branch 'ufs/dev' into add-qv-photochemistry-diagnostics

DWesl committed:
Merge remote-tracking branch 'origin/ufs/dev' into add-qv-photochemistry-diagnostics

grantfirl commented on physics/photochem/module_h2ophys.F90:
```diff
@@ -228,6 +231,9 @@ subroutine run(this, dt, p, h2opltc, h2o)
        enddo
+
+       if (present(dqv_dt_prd)) dqv_dt_prd(:, iLev) = pltc(:, 1) * dt
+       if (present(dqv_dt_qv))  dqv_dt_qv(:, iLev)  = (h2o(:, iLev) - pltc(:, 3)) * dt / pltc(:, 2)
```
@DWesl In debug mode tests, this line is leading to a runtime failure due to division by zero.

DWesl replied:
The code above does something like the below, I'm running regression tests on this variant now.
```suggestion
       if (present(dqv_dt_qv))  then
          where (pltc(:, 2) .ne. 0)
             dqv_dt_qv(:, iLev)  = (h2o(:, iLev) - pltc(:, 3)) * dt / pltc(:, 2)
          elsewhere
             dqv_dt_qv(:, iLev) = 0
          end where
       end if
```

DWesl replied:
Right, the `where` condition should do the division only if the denominator is *not* zero.  Edited above.

DWesl committed:
BUG: Avoid divide-by-zero

DWesl committed:
Merge branch 'ufs/dev' of ufs-community/ccpp-physics into add-qv-photochemistry-diagnostics

DWesl committed:
BUG: Allow photochemistry schemes to have parameters at different times.

In particular, allow the ozone scheme to have two extra sets of parameters during the ozone hole without crashing the model.

DWesl committed:
Merge branch 'ufs/dev' of ufs-community/ccpp-physics into photochemistry-index-fix

DWesl committed:
STY: Move the find_photochemistry_index calls inside the conditionals.

With a separate index for each photochemistry scheme, we can move the index-finding calls
in the conditional for the photochemistry scheme.

DWesl commented:
Given the talk of merging PRs on ufs-community/ufs-weather-model#2739, should I merge #286 into this one?

grantfirl commented:
> Given the talk of merging PRs on [ufs-community/ufs-weather-model#2739](https://github.com/ufs-community/ufs-weather-model/pull/2739), should I merge #286 into this one?

Yes, IMO, that will save some testing work.

DWesl committed:
Merge branch 'photochemistry-index-fix' into add-qv-photochemistry-diagnostics

DWesl committed:
Merge branch 'ufs/dev' of ufs-community/ccpp-physics into add-qv-photochemistry-diagnostics

DWesl committed:
Merge branch 'ufs/dev' of ufs-community/ccpp-physics into add-qv-photochemistry-diagnostics
